File: Android/app/src/main/java/ru/moonbaystudio/mympsu/ui/theme/AppTheme.kt
package ru.moonbaystudio.mympsu.ui.theme

import androidx.compose.ui.graphics.Color

enum class ThemedComponentShape {
    CAPSULE, ROUNDED_CARD, CLOUD
}

enum class ThemeBackgroundStyle {
    SYSTEM, DREAMY_SKY_DAY, DREAMY_SKY_NIGHT, TOKYO_CITY_DAY, TOKYO_CITY_NIGHT, GIRLY_VIBES_DAY, GIRLY_VIBES_NIGHT, CHUCK_NETWORK_DAY, CHUCK_NETWORK_NIGHT
}

data class AppTheme(
    val id: String,
    val name: String,
    val subtitle: String,
    val headerShape: ThemedComponentShape,
    val dateShape: ThemedComponentShape,
    val inputShape: ThemedComponentShape,
    val backgroundStyle: ThemeBackgroundStyle,
    val usesCloudPattern: Boolean,
    val isDark: Boolean,
    val primaryColor: Color
) {
    val usesCloudSurface: Boolean
        get() = headerShape == ThemedComponentShape.CLOUD || dateShape == ThemedComponentShape.CLOUD || usesCloudPattern

    val cloudSurfaceFill: Color
        get() = if (backgroundStyle == ThemeBackgroundStyle.DREAMY_SKY_NIGHT) {
            Color.Black.copy(alpha = 0.18f)
        } else {
            Color.White.copy(alpha = 0.26f)
        }

    val cloudSurfaceStroke: Color
        get() = if (backgroundStyle == ThemeBackgroundStyle.DREAMY_SKY_NIGHT) {
            Color.White.copy(alpha = 0.07f)
        } else {
            Color.White.copy(alpha = 0.34f)
        }
}

data class AppThemeFamily(
    val id: String,
    val name: String,
    val subtitle: String,
    val lightThemeID: String,
    val darkThemeID: String
)

object AppThemeCatalog {
    val families = listOf(
        AppThemeFamily("standard", "Стандартная", "Спокойный системный стиль без декоративного фона.", "classic", "classic-dark"),
        AppThemeFamily("clouds", "Облачно", "Мягкое небо, облака и легкая воздушная подложка.", "dreamy-clouds-day", "dreamy-clouds-night"),
        AppThemeFamily("tokyo-city", "Токио Сити", "Городской неон, стекло и холодные контрастные акценты.", "tokyo-city-day", "tokyo-city-night"),
        AppThemeFamily("girly-vibes", "Девчачие вайбы", "Розовые, ягодные и лавандовые оттенки с мягким светом.", "girly-vibes-day", "girly-vibes-night"),
        AppThemeFamily("chuck-network", "Чак нетворк", "Неон, сетка и энергичная цифровая атмосфера.", "chuck-network-day", "chuck-network-night")
    )

    val themes = listOf(
        AppTheme("classic", "Стандартная • Светлая", "Системная светлая тема", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.SYSTEM, false, false, Purple40),
        AppTheme("classic-dark", "Стандартная • Темная", "Системная темная тема", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.SYSTEM, false, true, Purple80),
        AppTheme("dreamy-clouds-day", "Облачно • Светлая", "Небо и мягкие облака", ThemedComponentShape.CLOUD, ThemedComponentShape.CLOUD, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.DREAMY_SKY_DAY, true, false, Color(0xFF007AFF)),
        AppTheme("dreamy-clouds-night", "Облачно • Темная", "Ночное небо, звезды и темные облака", ThemedComponentShape.CLOUD, ThemedComponentShape.CLOUD, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.DREAMY_SKY_NIGHT, true, true, Color(0xFF5AC8FA)),
        AppTheme("tokyo-city-day", "Токио Сити • Светлая", "Светлый городской неон", ThemedComponentShape.ROUNDED_CARD, ThemedComponentShape.ROUNDED_CARD, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.TOKYO_CITY_DAY, false, false, TokyoDayAccent),
        AppTheme("tokyo-city-night", "Токио Сити • Темная", "Ночной город и неоновые линии", ThemedComponentShape.ROUNDED_CARD, ThemedComponentShape.ROUNDED_CARD, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.TOKYO_CITY_NIGHT, false, true, TokyoNightAccent),
        AppTheme("girly-vibes-day", "Девчачие вайбы • Светлая", "Розовый свет и лавандовая база", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.GIRLY_VIBES_DAY, false, false, GirlyDayAccent),
        AppTheme("girly-vibes-night", "Девчачие вайбы • Темная", "Ягодный градиент и мягкий неон", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.GIRLY_VIBES_NIGHT, false, true, GirlyNightAccent),
        AppTheme("chuck-network-day", "Чак нетворк • Светлая", "Цифровая сетка и кислотный акцент", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.CHUCK_NETWORK_DAY, false, false, ChuckDayAccent),
        AppTheme("chuck-network-night", "Чак нетворк • Темная", "Черный неон и сетевой glow", ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemedComponentShape.CAPSULE, ThemeBackgroundStyle.CHUCK_NETWORK_NIGHT, false, true, ChuckNightAccent)
    )

    fun theme(id: String): AppTheme = themes.find { it.id == id } ?: themes[0]
}

File: iOS/MyMPSU/Models/AppTheme.swift
import SwiftUI

enum ThemedComponentShape: String, Codable, CaseIterable, Identifiable {
    case capsule
    case roundedCard
    case cloud

    var id: String { rawValue }
}

enum AppThemeBackgroundStyle: String, Codable, CaseIterable {
    case system
    case dreamySkyDay
    case dreamySkyNight
    case tokyoCityDay
    case tokyoCityNight
    case girlyVibesDay
    case girlyVibesNight
    case chuckNetworkDay
    case chuckNetworkNight
}

struct AppTheme: Identifiable, Hashable {
    let id: String
    let name: String
    let backgroundStyle: AppThemeBackgroundStyle
    let headerShape: ThemedComponentShape
    let inputShape: ThemedComponentShape
    let preferredColorScheme: ColorScheme?
    let usesCloudSurface: Bool
    let usesCloudPattern: Bool
    let cloudSurfaceFill: Color
    let cloudSurfaceStroke: Color

    var dateShape: ThemedComponentShape { headerShape }

    static func == (lhs: AppTheme, rhs: AppTheme) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

struct AppThemeFamily: Identifiable, Hashable {
    let id: String
    let name: String
    let subtitle: String
    let lightThemeID: String
    let darkThemeID: String
}

enum AppThemeCatalog {
    static let `default` = "system-light"

    static let families: [AppThemeFamily] = [
        AppThemeFamily(
            id: "system",
            name: "Системная",
            subtitle: "Чистый стандартный интерфейс.",
            lightThemeID: "system-light",
            darkThemeID: "system-dark"
        ),
        AppThemeFamily(
            id: "dreamy-sky",
            name: "Dreamy Sky",
            subtitle: "Мягкое небо и облачные поверхности.",
            lightThemeID: "dreamy-sky-day",
            darkThemeID: "dreamy-sky-night"
        ),
        AppThemeFamily(
            id: "tokyo-city",
            name: "Tokyo City",
            subtitle: "Городской неон для светлой и тёмной темы.",
            lightThemeID: "tokyo-city-day",
            darkThemeID: "tokyo-city-night"
        ),
        AppThemeFamily(
            id: "girly-vibes",
            name: "Girly Vibes",
            subtitle: "Мягкие розовые и лавандовые тона.",
            lightThemeID: "girly-vibes-day",
            darkThemeID: "girly-vibes-night"
        ),
        AppThemeFamily(
            id: "chuck-network",
            name: "Chuck Network",
            subtitle: "Свежая зелёная сетка и контраст.",
            lightThemeID: "chuck-network-day",
            darkThemeID: "chuck-network-night"
        )
    ]

    private static let themes: [String: AppTheme] = {
        let all: [AppTheme] = [
            AppTheme(
                id: "system-light",
                name: "Системная светлая",
                backgroundStyle: .system,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .light,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.78),
                cloudSurfaceStroke: Color.white.opacity(0.28)
            ),
            AppTheme(
                id: "system-dark",
                name: "Системная тёмная",
                backgroundStyle: .system,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .dark,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.12),
                cloudSurfaceStroke: Color.white.opacity(0.18)
            ),
            AppTheme(
                id: "dreamy-sky-day",
                name: "Dreamy Sky Day",
                backgroundStyle: .dreamySkyDay,
                headerShape: .cloud,
                inputShape: .roundedCard,
                preferredColorScheme: .light,
                usesCloudSurface: true,
                usesCloudPattern: true,
                cloudSurfaceFill: Color.white.opacity(0.78),
                cloudSurfaceStroke: Color.white.opacity(0.48)
            ),
            AppTheme(
                id: "dreamy-sky-night",
                name: "Dreamy Sky Night",
                backgroundStyle: .dreamySkyNight,
                headerShape: .cloud,
                inputShape: .roundedCard,
                preferredColorScheme: .dark,
                usesCloudSurface: true,
                usesCloudPattern: true,
                cloudSurfaceFill: Color.white.opacity(0.13),
                cloudSurfaceStroke: Color.white.opacity(0.20)
            ),
            AppTheme(
                id: "tokyo-city-day",
                name: "Tokyo City Day",
                backgroundStyle: .tokyoCityDay,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .light,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.70),
                cloudSurfaceStroke: Color.white.opacity(0.28)
            ),
            AppTheme(
                id: "tokyo-city-night",
                name: "Tokyo City Night",
                backgroundStyle: .tokyoCityNight,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .dark,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.12),
                cloudSurfaceStroke: Color.white.opacity(0.20)
            ),
            AppTheme(
                id: "girly-vibes-day",
                name: "Girly Vibes Day",
                backgroundStyle: .girlyVibesDay,
                headerShape: .roundedCard,
                inputShape: .roundedCard,
                preferredColorScheme: .light,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.72),
                cloudSurfaceStroke: Color.white.opacity(0.32)
            ),
            AppTheme(
                id: "girly-vibes-night",
                name: "Girly Vibes Night",
                backgroundStyle: .girlyVibesNight,
                headerShape: .roundedCard,
                inputShape: .roundedCard,
                preferredColorScheme: .dark,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.14),
                cloudSurfaceStroke: Color.white.opacity(0.20)
            ),
            AppTheme(
                id: "chuck-network-day",
                name: "Chuck Network Day",
                backgroundStyle: .chuckNetworkDay,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .light,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.70),
                cloudSurfaceStroke: Color.white.opacity(0.30)
            ),
            AppTheme(
                id: "chuck-network-night",
                name: "Chuck Network Night",
                backgroundStyle: .chuckNetworkNight,
                headerShape: .capsule,
                inputShape: .roundedCard,
                preferredColorScheme: .dark,
                usesCloudSurface: false,
                usesCloudPattern: false,
                cloudSurfaceFill: Color.white.opacity(0.12),
                cloudSurfaceStroke: Color.white.opacity(0.20)
            )
        ]
        return Dictionary(uniqueKeysWithValues: all.map { ($0.id, $0) })
    }()

    static func theme(for id: String) -> AppTheme {
        themes[id] ?? themes[Self.default]!
    }
}
